File: easy/94_binary_tree_inorder_traversal.rb
# 94. Binary Tree Inorder Traversal
# Easy
#
# Given the root of a binary tree, return the inorder traversal of its nodes'
# values.

class TreeNode
  attr_accessor :val, :left, :right

  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

def inorder_traversal(root)
  return [] if root.nil?

  inorder_traversal(root.left).concat([root.val], inorder_traversal(root.right))
end

File: go/easy/94_binary_tree_inorder_traversal.go
// 94. Binary Tree Inorder Traversal
// Easy
//
// Given the root of a binary tree, return the inorder traversal of its nodes'
// values.
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func inorderTraversal(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	var left []int
	if root.Left != nil {
		left = inorderTraversal(root.Left)
	}

	var right []int
	if root.Right != nil {
		right = inorderTraversal(root.Right)
	}

	return append(left, append([]int{root.Val}, right...)...)
}
